library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
             }
 
             if (setState(PINCHING)) {
-                val eps = 0.0001f
-                if (Math.abs(mAbsTargetX) < eps || Math.abs(mAbsTargetY) < eps) {
-                    // We want to interpret this as a scaled value, to work with the *actual* zoom.
-                    @ScaledPan var scaledFocusX = -detector.focusX
-                    @ScaledPan var scaledFocusY = -detector.focusY
-                    LOG.i("onScale:", "Setting focus.", "detectorFocusX:", scaledFocusX, "detectorFocusX:", scaledFocusY)
-
-                    // Account for current pan.
-                    scaledFocusX += scaledPanX
-                    scaledFocusY += scaledPanY
-
+                // We want to interpret this as a scaled value, to work with the *actual* zoom.
+                scaledFocusX = -detector.focusX
+                scaledFocusY = -detector.focusY
+                LOG.i("onScale:", "Setting focus.", "detectorFocusX:", scaledFocusX, "detectorFocusX:", scaledFocusY)
+
+                // Account for current pan.
+                scaledFocusX += scaledPanX
+                scaledFocusY += scaledPanY
+
+                mCurrentAbsTargetX = panX
+                mCurrentAbsTargetY = panY
+                if (mInitialAbsTargetX.isNaN() || mInitialAbsTargetY.isNaN()) {
                     // Transform to an absolute, scale-independent value.
-                    mAbsTargetX = unresolvePan(scaledFocusX)
-                    mAbsTargetY = unresolvePan(scaledFocusY)
-                    LOG.i("onScale:", "Setting focus.", "absTargetX:", mAbsTargetX, "absTargetY:", mAbsTargetY)
+                    mInitialAbsTargetX = unresolvePan(scaledFocusX)
+                    mInitialAbsTargetY = unresolvePan(scaledFocusY)
+                    LOG.i("onScale:", "Setting initial focus.", "absTargetX:", mInitialAbsTargetX, "absTargetY:", mInitialAbsTargetY)
+                } else {
+                    // when the initial focus point is set, use it to
+                    // calculate the location difference to the current focus point
+                    mCurrentAbsTargetX += mInitialAbsTargetX - unresolvePan(scaledFocusX)
+                    mCurrentAbsTargetY += mInitialAbsTargetY - unresolvePan(scaledFocusY)
                 }
 
                 // Having both overPinch and overScroll is hard to manage, there are lots of bugs if we do.
                 val factor = detector.scaleFactor
                 val newZoom = zoom * factor
-                applyPinch(newZoom, mAbsTargetX, mAbsTargetY, true)
+
+                // apply the new zoom and pan values accordingly
+                applyZoomAndAbsolutePan(newZoom,
+                        mCurrentAbsTargetX, mCurrentAbsTargetY,
```

Comment:
I just realized this pinch target is wrong since it should be a value in view bounds, not an absolute pan value